In [1]:
import pandas as pd

In [2]:
from google.colab import files

uploaded = files.upload()

Saving Plasma_Metabolites_In_PPMS.csv to Plasma_Metabolites_In_PPMS.csv


In [3]:
df = pd.read_csv('Plasma_Metabolites_In_PPMS.csv', header=1)

Dropping all columns with Not a Number in them (NaN)

In [4]:
df

Unnamed: 0                            proposed Metabolite  \
0           NaN                                   Trigonelline   
1           NaN                   N-[(E,E)-Piperoyl]piperidine   
2           NaN                       N3-hydroxypropylcytosine   
3           NaN           27alpha-Hydroxy-8-dehydrocholesterol   
4           NaN  6-amino-5[N-methylformylamino]-1-methyluracil   
..          ...                                            ...   
529         NaN                         2-Methylbutyrylglycine   
530         NaN                    trans-2-Dodecenoylcarnitine   
531         NaN                          N-palmitoyl threonine   
532         NaN                                     L-Urobilin   
533         NaN                                 Palmitoleamide   

    proposed Formula    Mass m/z   ppm Polarity Adduct  RSD_Control_A [%]  \
0            C7H7NO2  137.047805  0.92      pos      H              90.30   
1          C17H19NO3  285.136699  0.72      pos      H             103.37   
2          C7H11N3O2  169.085107 -0.12      pos      H             143.24   
3           C27H44O2  400.333647 -1.21      pos      H              42.15   
4          C7H10N4O3  198.075500  1.06      pos      H              76.69   
..               ...         ...   ...      ...    ...                ...   
529         C7H13NO3  159.089549  0.03      pos      H              62.72   
530        C19H35NO4  341.256986  1.11      pos      H              30.16   
531        C20H39NO4  357.287525 -1.07      pos      H              33.93   
532       C33H46N4O6  594.340045 -2.84      pos      H              40.97   
533         C16H31NO  253.240496 -0.27      pos      H              29.87   

     RSD_Control_B[%]  RSD_PPMS_A[%]  ...  PPMS_B_P11  PPMS_B_P12  PPMS_B_P13  \
0               84.85         164.47  ...     94900.0    529000.0   1210000.0   
1              118.45         148.59  ...     43500.0      3880.0    491000.0   
2              112.00         124.89  ...    377000.0     62500.0   1060000.0   
3               65.87          54.52  ...     10200.0     11000.0     11800.0   
4               46.31          53.35  ...     18000.0     32900.0     11800.0   
..                ...            ...  ...         ...         ...         ...   
529             47.01          92.57  ...      8220.0      7880.0     11500.0   
530             51.28          55.90  ...     44700.0     25500.0     29100.0   
531             76.60         108.63  ...      3070.0     24000.0     21000.0   
532             98.08          84.79  ...     11600.0      2640.0      5890.0   
533            165.24          82.49  ...      8490.0     15100.0      6180.0   

     PPMS_B_P14  PPMS_B_P15  PPMS_B_P16  PPMS_B_P17  PPMS_B_P18  PPMS_B_P19  \
0      502000.0     12300.0    208000.0   2290000.0   2170000.0   1400000.0   
1      106000.0     43400.0     83700.0    198000.0     47800.0    506000.0   
2      313000.0    325000.0    371000.0    167000.0     43900.0    319000.0   
3       13200.0     25700.0      6250.0     16200.0      9490.0     10600.0   
4       21100.0      7950.0     31400.0    102000.0     22300.0     36900.0   
..          ...         ...         ...         ...         ...         ...   
529      6680.0      5280.0      4570.0      4160.0      4850.0      5270.0   
530     22000.0     44200.0     49000.0     38700.0     45200.0     64600.0   
531     10800.0     10800.0      8740.0      6820.0      7130.0     19600.0   
532      7000.0     25600.0      5050.0      7720.0      8510.0     15800.0   
533      6930.0      5560.0      9760.0      9960.0      8820.0      7690.0   

     PPMS_B_P20  
0       78200.0  
1      214000.0  
2      111000.0  
3        8240.0  
4        4790.0  
..          ...  
529     13200.0  
530     21000.0  
531      6040.0  
532     63200.0  
533      6090.0  

[534 rows x 93 columns]

In [13]:
df.dropna(axis=1)

proposed Metabolite proposed Formula  \
0                                     Trigonelline          C7H7NO2   
1                     N-[(E,E)-Piperoyl]piperidine        C17H19NO3   
2                         N3-hydroxypropylcytosine        C7H11N3O2   
3             27alpha-Hydroxy-8-dehydrocholesterol         C27H44O2   
4    6-amino-5[N-methylformylamino]-1-methyluracil        C7H10N4O3   
..                                             ...              ...   
529                         2-Methylbutyrylglycine         C7H13NO3   
530                    trans-2-Dodecenoylcarnitine        C19H35NO4   
531                          N-palmitoyl threonine        C20H39NO4   
532                                     L-Urobilin       C33H46N4O6   
533                                 Palmitoleamide         C16H31NO   

       Mass m/z   ppm Polarity Adduct  RSD_Control_A [%]  RSD_Control_B[%]  \
0    137.047805  0.92      pos      H              90.30             84.85   
1    285.136699  0.72      pos      H             103.37            118.45   
2    169.085107 -0.12      pos      H             143.24            112.00   
3    400.333647 -1.21      pos      H              42.15             65.87   
4    198.075500  1.06      pos      H              76.69             46.31   
..          ...   ...      ...    ...                ...               ...   
529  159.089549  0.03      pos      H              62.72             47.01   
530  341.256986  1.11      pos      H              30.16             51.28   
531  357.287525 -1.07      pos      H              33.93             76.60   
532  594.340045 -2.84      pos      H              40.97             98.08   
533  253.240496 -0.27      pos      H              29.87            165.24   

     RSD_PPMS_A[%]  RSD_PPMS_B[%]  ...  PPMS_B_P11  PPMS_B_P12  PPMS_B_P13  \
0           164.47          94.17  ...     94900.0    529000.0   1210000.0   
1           148.59         110.54  ...     43500.0      3880.0    491000.0   
2           124.89          82.15  ...    377000.0     62500.0   1060000.0   
3            54.52          61.06  ...     10200.0     11000.0     11800.0   
4            53.35          78.14  ...     18000.0     32900.0     11800.0   
..             ...            ...  ...         ...         ...         ...   
529          92.57          75.28  ...      8220.0      7880.0     11500.0   
530          55.90          48.25  ...     44700.0     25500.0     29100.0   
531         108.63          60.95  ...      3070.0     24000.0     21000.0   
532          84.79         100.66  ...     11600.0      2640.0      5890.0   
533          82.49         152.44  ...      8490.0     15100.0      6180.0   

     PPMS_B_P14  PPMS_B_P15  PPMS_B_P16  PPMS_B_P17  PPMS_B_P18  PPMS_B_P19  \
0      502000.0     12300.0    208000.0   2290000.0   2170000.0   1400000.0   
1      106000.0     43400.0     83700.0    198000.0     47800.0    506000.0   
2      313000.0    325000.0    371000.0    167000.0     43900.0    319000.0   
3       13200.0     25700.0      6250.0     16200.0      9490.0     10600.0   
4       21100.0      7950.0     31400.0    102000.0     22300.0     36900.0   
..          ...         ...         ...         ...         ...         ...   
529      6680.0      5280.0      4570.0      4160.0      4850.0      5270.0   
530     22000.0     44200.0     49000.0     38700.0     45200.0     64600.0   
531     10800.0     10800.0      8740.0      6820.0      7130.0     19600.0   
532      7000.0     25600.0      5050.0      7720.0      8510.0     15800.0   
533      6930.0      5560.0      9760.0      9960.0      8820.0      7690.0   

     PPMS_B_P20  
0       78200.0  
1      214000.0  
2      111000.0  
3        8240.0  
4        4790.0  
..          ...  
529     13200.0  
530     21000.0  
531      6040.0  
532     63200.0  
533      6090.0  

[534 rows x 92 columns]

In [50]:
PPMS_Individual_Metabolites = df.drop(columns=["proposed Metabolite", "proposed Formula", "Mass m/z", "ppm", "Polarity", "Adduct","RSD_Control_A [%]","RSD_Control_B[%]", "RSD_PPMS_A[%]", "RSD_PPMS_B[%]","VIP_comp_1_A", "VIP_comp_1_A rank", "VIP_comp_1_B", "VIP_comp_1_B rank", "VIP_comp_2_A", "VIP_comp_2_A rank", "VIP_comp_2_B", "VIP_comp_2_B rank", "Weighted VIP score A", "Weighted VIP score A rank", "Weighted VIP score B", "Weighted VIP score B rank", "mean Intensity Control A", "mean Intensity PPMS A", "mean Intensity Control B", "mean Intensity PPMS B"])

Dropping all columns beside the individual values of the metabolites for the controls and PPMS patients. This will make up the training data. I could not use the mean intensities as that would provide too small of a sample.

In [51]:
PPMS_Individual_Metabolites

Control_A_P1  Control_A_P2  Control_A_P3  Control_A_P4  Control_A_P5  \
0        662000.0     1790000.0      282000.0     3190000.0     3290000.0   
1       1210000.0      168000.0      517000.0      118000.0       84600.0   
2         47000.0      334000.0      138000.0       43700.0      317000.0   
3         16800.0        6020.0       10900.0       13600.0       12500.0   
4         23800.0       23000.0       12100.0       40100.0       54600.0   
..            ...           ...           ...           ...           ...   
529        3470.0       10300.0        2450.0        4120.0        7170.0   
530       24600.0       35800.0       17800.0       43200.0       35700.0   
531        7750.0        9590.0        6100.0        8680.0        7170.0   
532        2620.0        7320.0        8920.0        3220.0       14400.0   
533        3910.0        5890.0        2920.0        7910.0        8980.0   

     Control_A_P6  Control_A_P7  Control_A_P8  Control_A_P9  Control_A_P10  \
0        137000.0     2910000.0       24600.0     1520000.0       320000.0   
1        470000.0       59200.0      548000.0       38000.0       295000.0   
2        142000.0       41300.0     1270000.0       36000.0        91400.0   
3         14100.0       24100.0       27200.0       13600.0        14000.0   
4          8680.0       86200.0        6580.0       65900.0        11700.0   
..            ...           ...           ...           ...            ...   
529        6790.0       18300.0        3580.0        6460.0         3310.0   
530       23200.0       15900.0       35000.0       30900.0        51700.0   
531        4930.0        2340.0       11500.0        4670.0         6950.0   
532        9190.0        4830.0        7100.0       13700.0         9410.0   
533        4930.0        8790.0        7530.0        3880.0         5670.0   

     ...  PPMS_B_P11  PPMS_B_P12  PPMS_B_P13  PPMS_B_P14  PPMS_B_P15  \
0    ...     94900.0    529000.0   1210000.0    502000.0     12300.0   
1    ...     43500.0      3880.0    491000.0    106000.0     43400.0   
2    ...    377000.0     62500.0   1060000.0    313000.0    325000.0   
3    ...     10200.0     11000.0     11800.0     13200.0     25700.0   
4    ...     18000.0     32900.0     11800.0     21100.0      7950.0   
..   ...         ...         ...         ...         ...         ...   
529  ...      8220.0      7880.0     11500.0      6680.0      5280.0   
530  ...     44700.0     25500.0     29100.0     22000.0     44200.0   
531  ...      3070.0     24000.0     21000.0     10800.0     10800.0   
532  ...     11600.0      2640.0      5890.0      7000.0     25600.0   
533  ...      8490.0     15100.0      6180.0      6930.0      5560.0   

     PPMS_B_P16  PPMS_B_P17  PPMS_B_P18  PPMS_B_P19  PPMS_B_P20  
0      208000.0   2290000.0   2170000.0   1400000.0     78200.0  
1       83700.0    198000.0     47800.0    506000.0    214000.0  
2      371000.0    167000.0     43900.0    319000.0    111000.0  
3        6250.0     16200.0      9490.0     10600.0      8240.0  
4       31400.0    102000.0     22300.0     36900.0      4790.0  
..          ...         ...         ...         ...         ...  
529      4570.0      4160.0      4850.0      5270.0     13200.0  
530     49000.0     38700.0     45200.0     64600.0     21000.0  
531      8740.0      6820.0      7130.0     19600.0      6040.0  
532      5050.0      7720.0      8510.0     15800.0     63200.0  
533      9760.0      9960.0      8820.0      7690.0      6090.0  

[534 rows x 66 columns]

In [52]:
PPMS_Individual_Metabolites_Transposed = PPMS_Individual_Metabolites.T

Swapping the Columns and Rows for the machine learning model to read the dataframe properly

In [53]:
PPMS_Individual_Metabolites_Transposed

0          1         2        3         4        5    \
Control_A_P1   662000.0  1210000.0   47000.0  16800.0   23800.0   9270.0   
Control_A_P2  1790000.0   168000.0  334000.0   6020.0   23000.0  14700.0   
Control_A_P3   282000.0   517000.0  138000.0  10900.0   12100.0  15200.0   
Control_A_P4  3190000.0   118000.0   43700.0  13600.0   40100.0  10000.0   
Control_A_P5  3290000.0    84600.0  317000.0  12500.0   54600.0  16400.0   
...                 ...        ...       ...      ...       ...      ...   
PPMS_B_P16     208000.0    83700.0  371000.0   6250.0   31400.0   8480.0   
PPMS_B_P17    2290000.0   198000.0  167000.0  16200.0  102000.0   9760.0   
PPMS_B_P18    2170000.0    47800.0   43900.0   9490.0   22300.0  12100.0   
PPMS_B_P19    1400000.0   506000.0  319000.0  10600.0   36900.0   9590.0   
PPMS_B_P20      78200.0   214000.0  111000.0   8240.0    4790.0   8820.0   

                  6        7        8         9    ...      524      525  \
Control_A_P1  12000.0   7290.0  57600.0  113000.0  ...  22900.0  43300.0   
Control_A_P2  15800.0  14200.0  77200.0  146000.0  ...  19800.0  42400.0   
Control_A_P3  16000.0   2900.0  61500.0   74800.0  ...  28800.0  47200.0   
Control_A_P4   6460.0  12000.0  51700.0  108000.0  ...  16100.0  68400.0   
Control_A_P5  36900.0   4280.0  37300.0   76100.0  ...  45400.0  23400.0   
...               ...      ...      ...       ...  ...      ...      ...   
PPMS_B_P16    14300.0   8590.0  42800.0   71200.0  ...  32400.0  67600.0   
PPMS_B_P17    17700.0  11600.0  43900.0   71200.0  ...  28800.0  27100.0   
PPMS_B_P18    14700.0   5750.0  15600.0   28700.0  ...  20900.0  25300.0   
PPMS_B_P19    30900.0   7320.0  24100.0   34100.0  ...  41700.0  23500.0   
PPMS_B_P20    15700.0   6470.0  91200.0  132000.0  ...  16900.0  38900.0   

                   526        527       528      529      530      531  \
Control_A_P1   61000.0   458000.0  104000.0   3470.0  24600.0   7750.0   
Control_A_P2  118000.0   646000.0  167000.0  10300.0  35800.0   9590.0   
Control_A_P3   50300.0   641000.0   88200.0   2450.0  17800.0   6100.0   
Control_A_P4   64300.0   681000.0  132000.0   4120.0  43200.0   8680.0   
Control_A_P5  175000.0   724000.0  202000.0   7170.0  35700.0   7170.0   
...                ...        ...       ...      ...      ...      ...   
PPMS_B_P16    125000.0  1080000.0  201000.0   4570.0  49000.0   8740.0   
PPMS_B_P17    101000.0  1290000.0  148000.0   4160.0  38700.0   6820.0   
PPMS_B_P18     87500.0   647000.0  143000.0   4850.0  45200.0   7130.0   
PPMS_B_P19    122000.0   922000.0  193000.0   5270.0  64600.0  19600.0   
PPMS_B_P20     59100.0   596000.0   89100.0  13200.0  21000.0   6040.0   

                  532     533  
Control_A_P1   2620.0  3910.0  
Control_A_P2   7320.0  5890.0  
Control_A_P3   8920.0  2920.0  
Control_A_P4   3220.0  7910.0  
Control_A_P5  14400.0  8980.0  
...               ...     ...  
PPMS_B_P16     5050.0  9760.0  
PPMS_B_P17     7720.0  9960.0  
PPMS_B_P18     8510.0  8820.0  
PPMS_B_P19    15800.0  7690.0  
PPMS_B_P20    63200.0  6090.0  

[66 rows x 534 columns]

In [54]:
labels = []
for col in PPMS_Individual_Metabolites_Transposed.index:
    if "Control" in col:
        labels.append(0)
    elif "PPMS" in col:
        labels.append(1)
    else:
        labels.append(None)

This "labels" function tells the program that whenever it sees the word Control, PPMS or anything else in the rows, it should label them either 0, 1 or nothing. This is so that when the Random Forest Classifier is run, it can do binary classification in a form that it understands.

In [55]:
feature_cols = [i for i in range(PPMS_Individual_Metabolites_Transposed.shape[1])]

The "feature_cols" variable is the way to isolate columns 0 to 533 in the dataframe so that it can be classified as X which is the form that the Random Forest Classifier needs to be able to perform functions based on the dataframe.

In [56]:
X = PPMS_Individual_Metabolites_Transposed[feature_cols]
y = labels

Classifying my variables for Random Forest Classification

In [58]:
import sklearn

In [78]:
from sklearn.ensemble import RandomForestClassifier
from sklearn.model_selection import train_test_split
from sklearn.metrics  import accuracy_score, confusion_matrix, classification_report

In [79]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

In [83]:
Data_Model = RandomForestClassifier(random_state = 42)

In [84]:
Data_Model.fit(X_train, y_train)

RandomForestClassifier(random_state=42)

In [85]:
Test_predictions = Data_Model.predict(X_test)

In [87]:
print("Accuracy:", accuracy_score(y_test, Test_predictions))
print("Confusion Matrix:", confusion_matrix(y_test, Test_predictions))
print("Classification Report", classification_report(y_test, Test_predictions))

Accuracy: 0.5714285714285714
Confusion Matrix: [[2 6]
 [0 6]]
Classification Report               precision    recall  f1-score   support

           0       1.00      0.25      0.40         8
           1       0.50      1.00      0.67         6

    accuracy                           0.57        14
   macro avg       0.75      0.62      0.53        14
weighted avg       0.79      0.57      0.51        14



Interpreting these metrics for the accuracy of my Random Forest Classifier:

Accuracy: As a percentage, this would be 57%. My model is correctly predicting whether a patient has PPMS or is a control only 57% percent of the time.

Confusion Matrix:

True Negatives - 2

False Positives - 6

False Negatives - 0

True Positives - 6

Out of the 14 predictions my model made, it successfully predicted 2 Controls (True Negatives) and 6 PPMS patients (True Positives). However, the model wrongly classified 6 Controls as PPMS patients (False Positives). Interestingly, there were no False Negatives.

Classification Report:
Recall = "When the truth is X, how often did I predict X?" -->

Precision = "When I predict X, how often am I right?"


For the negative labels (Controls), the precision was 100%. For the positive labels (PPMS patients), the precision was 50%. This means that whenever the model predicted that the patient was a control, it was correct 100% of the time. Whereas, whenever the model predicted the patient was a PPMS patient, it was correct only 50% of the time.

The recall value of the negative labels was 25%. The recall value of the positive labels was 100%. The model was not very good at detecting negatives and labelling them as such. But, it was very good at detecting positives.

In [89]:
from sklearn.model_selection import GridSearchCV

Because these initial results of my model's accuracy were not satisfactory, I am going to test what is the best amount of 'leaves' to use in my decision trees that make up my 'Random Forest'. This is to see the fit of my training data, but generally to improve accuracy. The 'leaves' or leaf nodes are the final point of a decision tree which is what the model uses to decide if a patient has PPMS or is a control based on the data points I have given it. In this context, I am training the model to predict if a patient has PPMS or not based on the concentration of 534 different metabolites in their blood plasma. So, one leaf node could be Metabolite 1 < 600 mM and Metabolite 236 > 500 mM means that patient has PPMS (Label = 1). A model can be overfit if the training data has too much variety - in this case, if a leaf node used the concentrations of 531 different metabolites to make a decision. Overfitting can lead to unreliable predictions on "unseen data". Underfitting would be the opposite. If your decision tree is too simple and only relies on a couple of variables to make a decision, then this can lead to the model not capturing important patterns in the data. One thing though is that Random Forest models do mitigate overfitting by using multiple "trees in the forest" by training on many random subsets.

In [90]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42, stratify=y)

In [99]:
Data_Model = RandomForestClassifier()

In [100]:
param_grid = {
    'max_depth': [3, 5, 7, 10, None],
    'max_leaf_nodes': [5, 10, 20, 50, None],
    'n_estimators': [100, 200],
    'random_state': [42]
}

In [101]:
grid_search = GridSearchCV(
    estimator=Data_Model,
    param_grid=param_grid,
    cv=5,
    scoring='accuracy',
    n_jobs=-1
)

In [103]:
grid_search.fit(X_train, y_train)

GridSearchCV(cv=5, estimator=RandomForestClassifier(), n_jobs=-1,
             param_grid={'max_depth': [3, 5, 7, 10, None],
                         'max_leaf_nodes': [5, 10, 20, 50, None],
                         'n_estimators': [100, 200], 'random_state': [42]},
             scoring='accuracy')

In [105]:
best_model = grid_search.best_estimator_

In [106]:
best_model

RandomForestClassifier(max_depth=3, max_leaf_nodes=10, random_state=42)

In [107]:
Test_predictions = best_model.predict(X_test)

In [109]:
print("Accuracy:", accuracy_score(y_test, Test_predictions))
print("Confusion Matrix:", confusion_matrix(y_test, Test_predictions))
print("Classification Report", classification_report(y_test, Test_predictions))

Accuracy: 0.7142857142857143
Confusion Matrix: [[6 1]
 [3 4]]
Classification Report               precision    recall  f1-score   support

           0       0.67      0.86      0.75         7
           1       0.80      0.57      0.67         7

    accuracy                           0.71        14
   macro avg       0.73      0.71      0.71        14
weighted avg       0.73      0.71      0.71        14



Here you can see that there has been a decent improvement in the weighted accuracy of my model, it is now correct about its predictions 71% of the time. This is still not excellent but this value could be for a variety of reasons:

Technical Reasons

- High feature to sample ratio. My data is 'high dimensional' meaning I have many features (Plasma metabolites) to represent one data point (The patient). This can lead to the 'curse of dimensionality' where the model's efficiency and effectiveness deteriorates as the dimensionality increases exponentially.
- Small data set. There were only 66 data points of PPMS and Healthy Control patients combined. This prevents the machine learning model from seeing patterns spanning over many examples. Capturing patterns is the most crucial part of machine learning.
- Feature selection. Using all of the metabolites in the data sheet from the paper may have resulted in less relevant metabolites being used for the training data. The paper does state that it used untargeted metabolomics in order to elucidate metabolic profiles specific to PPMS.

Biological Reasons
- Variability. The accepted  relative standard deviation of untargeted metabolomics is <10%. Due to my small data set, small amounts of biological noise and variability can cause my model to still overfit to any outliers. This point also highlights the importance of targeted metabolomics in disease diagnosis as well as the importance of continued research into disease biomarkers. The most common disease biomarker for PPMS is neurofilament light chain which has allowed for the reliable monitoring of disease progression and treatment response. However, there are still challenges in using disease biomarkers as a reliable way of diagnosing patients (Khalil, M., et al. (2024). 'Neurofilaments as biomarkers in neurological disorders — towards clinical application', *Nature Reviews Neurology*, 20, pp. 269–287. Available at: https://www.nature.com/articles/s41582-024-00955-x (Accessed September 10 2025).
- Are body fluid metabolites reliable? There are some challenges with using body fluids as a diagnostics tool for PPMS. One of which is that biomarkers in the blood or CSF lag behind the progression of PPMS itself. Meaning biomarker changes are only detected after untreatable neuronal damage (Biernacki, T., et al. (2022). 'Emerging Biomarkers of Multiple Sclerosis in the Blood and the CSF: A Focus on Neurofilaments and Therapeutic Considerations', *International Journal of Medical Sciences*, 23(6), pp. 3383. Available at: https://pmc.ncbi.nlm.nih.gov/articles/PMC8951485/ (Accessed September 10 2025)

Conclusion: There is lots of room for improvement in this project. My next steps will be to continuously learn about the intersection between big data in biology and AI as various fields move into using computational biology to solve big challenges that we will face in the future of medicine.